rl4co/envs/scheduling/fjsp/generator.py

```diff
@@ -0,0 +1,215 @@
+
+    """
+
+    def __init__(
+        self,
+        num_jobs: int = 10,
+        num_machines: int = 5,
+        min_ops_per_job: int = 4,
+        max_ops_per_job: int = 6,
+        min_processing_time: int = 1,
+        max_processing_time: int = 20,
+        min_eligible_ma_per_op: int = 1,
+        max_eligible_ma_per_op: int = None,
+        **unused_kwargs,
+    ):
+        self.num_jobs = num_jobs
+        self.num_mas = num_machines
+        self.min_ops_per_job = min_ops_per_job
+        self.max_ops_per_job = max_ops_per_job
+        self.min_processing_time = min_processing_time
+        self.max_processing_time = max_processing_time
+        self.min_eligible_ma_per_op = min_eligible_ma_per_op
+        self.max_eligible_ma_per_op = max_eligible_ma_per_op or num_machines
+        # determines whether to use a fixed number of total operations or let it vary between instances
+        # NOTE: due to the way rl4co builds datasets, we need a fixed size here
+        self.n_ops_max = max_ops_per_job * num_jobs
+
+        # FFSP environment doen't have any other kwargs
+        if len(unused_kwargs) > 0:
+            log.error(f"Found {len(unused_kwargs)} unused kwargs: {unused_kwargs}")
+
+    def _simulate_processing_times(
+        self, n_eligible_per_ops: torch.Tensor
+    ) -> torch.Tensor:
+        bs, n_ops_max = n_eligible_per_ops.shape
+
+        # (bs, max_ops, machines)
+        ma_seq_per_ops = torch.arange(1, self.num_mas + 1)[None, None].expand(
+            bs, n_ops_max, self.num_mas
+        )
+        # generate a matrix of size (ops, mas) per batch, each row having as many ones as the operation eligible machines
+        # E.g. n_eligible_per_ops=[1,3,2]; num_mas=4
+        # [[1,0,0,0],
+        #   1,1,1,0],
+        #   1,1,0,0]]
+        # This will be shuffled randomly to generate a machine-operation mapping
+        ma_ops_edges_unshuffled = torch.Tensor.float(
+            ma_seq_per_ops <= n_eligible_per_ops[..., None]
+        )
+        # random shuffling
+        idx = torch.rand_like(ma_ops_edges_unshuffled).argsort()
+        ma_ops_edges = ma_ops_edges_unshuffled.gather(2, idx).transpose(1, 2)
+
+        # (bs, max_ops, machines)
+        proc_times = torch.ones((bs, n_ops_max, self.num_mas))
+        proc_times = torch.randint(
+            self.min_processing_time,
+            self.max_processing_time + 1,
+            size=(bs, self.num_mas, n_ops_max),
+        )
+
+        # TODO: This should be unnecessary due to multiplication with ma-ops edges
+        # proc_times[mask[:,None].expand_as(proc_times)] = 0
+        # remove proc_times for which there is no corresponding ma-ops connection
+        proc_times = proc_times * ma_ops_edges
+        return proc_times
+
+    def _generate(self, batch_size) -> TensorDict:
+        batch_size = [batch_size] if isinstance(batch_size, int) else batch_size
```

Comment:
[Minor] This can also be removed since it is included in the base `Generator()` class

https://github.com/ai4co/rl4co/blob/684dbdcb6af83936a7733e814f1c9a42a280cab8/rl4co/envs/common/utils.py#L17-L19